samza-core/src/main/java/org/apache/samza/execution/StreamManager.java

```diff
@@ -75,4 +85,52 @@ public void createStreams(List<StreamSpec> streams) {
 
     return streamToPartitionCount;
   }
+
+  /**
+   * This is a best-effort approach to clear the internal streams from previous run, including intermediate streams,
+   * checkpoint stream and changelog streams.
+   * For batch processing, we always clean up the previous internal streams and create a new set for each run.
+   * @param prevConfig config of the previous run
+   */
+  public void clearStreamsFromPreviousRun(Config prevConfig) {
+    try {
+      ApplicationConfig appConfig = new ApplicationConfig(prevConfig);
+      LOGGER.info("run.id from previous run is {}", appConfig.getRunId());
+
+      StreamConfig streamConfig = new StreamConfig(prevConfig);
+
+      //Find all intermediate streams and clean up
+      Set<StreamSpec> intStreams = JavaConversions.asJavaCollection(streamConfig.getStreamIds()).stream()
+          .filter(streamConfig::getIsIntermediate)
+          .map(id -> new StreamSpec(id, streamConfig.getPhysicalName(id), streamConfig.getSystem(id)))
+          .collect(Collectors.toSet());
+      intStreams.forEach(stream -> {
+          LOGGER.info("Clear intermediate stream {} in system {}", stream.getPhysicalName(), stream.getSystemName());
+          sysAdmins.get(stream.getSystemName()).clearStream(stream);
+        });
+
+      //Find checkpoint stream and clean up
+      TaskConfig taskConfig = new TaskConfig(prevConfig);
+      String checkpointManagerFactoryClass = taskConfig.getCheckpointManagerFactory().getOrElse(defaultValue(null));
+      if (checkpointManagerFactoryClass != null) {
+        CheckpointManager checkpointManager = ((CheckpointManagerFactory) Util.getObj(checkpointManagerFactoryClass))
+            .getCheckpointManager(prevConfig, new MetricsRegistryMap());
+        checkpointManager.clearCheckpoints();
+      }
+
+      //Find changelog streams and remove them
+      StorageConfig storageConfig = new StorageConfig(prevConfig);
+      for (String store : JavaConversions.asJavaCollection(storageConfig.getStoreNames())) {
+        String changelog = storageConfig.getChangelogStream(store).getOrElse(defaultValue(null));
+        if (changelog != null) {
+          LOGGER.info("Clear store {} changelog {}", store, changelog);
+          SystemStream systemStream = Util.getSystemStreamFromNames(changelog);
+          StreamSpec spec = StreamSpec.createChangeLogStreamSpec(systemStream.getStream(), systemStream.getSystem(), 1);
+          sysAdmins.get(spec.getSystemName()).clearStream(spec);
+        }
+      }
+    } catch (Exception e) {
+      LOGGER.warn("Fail to clear internal streams from previous run. Please clean up manually.", e);
```

Comment:
I think we should rethrow here, since we can't guarantee correctness. Thoughts?

Reply:
this should be fine. We always generate a new set of topics (appending the run.id so they are different from previous runs). I actually expect clearStreams not supported in many scenarios, where we might rely on retention to get the streams cleared.

Reply:
Ahh, forgot about that. 👍 

So this is just to free up resources rather than to ensure correctness. I'd suggest mentioning that in a comment.